src/domain/auth/auth.repository.ts

```diff
@@ -1,26 +1,155 @@
+
+    // Store nonce for reference to verify/prevent replay attacks
+    const cacheDir = CacheRouter.getAuthNonceCacheDir(nonce);
+    await this.cacheService.set(cacheDir, nonce, this.nonceTtlInSeconds);
+
+    return {
+      nonce,
+    };
+  }
+
+  /**
+   * Verifies the validity of a signed message and returns an access token:
+   *
+   * 1. Ensure the message itself has not expired.
+   * 2. Ensure the nonce was generated by us/is not a replay attack.
+   * 3. Verify the signature of the message.
+   * 4. Return an access token if all checks pass.
+   *
+   * @param args - DTO containing the message and signature to verify.
+   *
+   * The following adhere to JWT standard {@link https://datatracker.ietf.org/doc/html/rfc7519}
+   *
+   * @returns accessToken - JWT access token
+   * @returns tokenType - token type ('Bearer') to be used in the `Authorization` header
+   * @returns notBefore - epoch from when token is valid (if applicable, otherwise null)
+   * @returns expiresIn - time in seconds until the token expires (if applicable, otherwise null)
+   */
+  async verify(args: VerifyAuthMessageDto): Promise<{
+    accessToken: string;
+    tokenType: string;
+    notBefore: number | null;
+    expiresIn: number | null;
+  }> {
+    const isAuthorized = await this.isAuthorized(args).catch(() => false);
 
-  generateNonce(): string {
-    return this.authApi.generateNonce();
+    if (!isAuthorized) {
+      throw new UnauthorizedException();
+    }
+
+    const dateWhenTokenIsValid = args.message.notBefore
+      ? new Date(args.message.notBefore)
+      : null;
+    const dateWhenTokenExpires = args.message.expirationTime
+      ? new Date(args.message.expirationTime)
+      : null;
+
+    const secondsUntilTokenIsValid = dateWhenTokenIsValid
+      ? this.getSecondsUntil(dateWhenTokenIsValid)
+      : null;
+    const secondsUntilTokenExpires = dateWhenTokenExpires
+      ? this.getSecondsUntil(dateWhenTokenExpires)
+      : null;
+
+    const accessToken = this.jwtService.sign(args.message, {
+      ...(secondsUntilTokenIsValid !== null && {
+        notBefore: secondsUntilTokenIsValid,
+      }),
+      ...(secondsUntilTokenExpires !== null && {
+        expiresIn: secondsUntilTokenExpires,
+      }),
+    });
+
+    return {
+      tokenType: AuthService.AUTH_TOKEN_TOKEN_TYPE,
+      accessToken,
+      // Differing measurements match JWT standard {@link https://datatracker.ietf.org/doc/html/rfc7519}
+      notBefore: dateWhenTokenIsValid?.getTime() ?? null,
+      expiresIn: secondsUntilTokenExpires,
+    };
   }
 
-  async verifyMessage(args: {
-    address: `0x${string}`;
+  /**
+   * Verifies that a message is valid according to its expiration date,
+   * signature and nonce.
+   *
+   * @param args.message - SiWe message in object form
+   * @param args.signature - signature from signing the message
+   *
+   * @returns boolean - whether the message is valid
+   */
+  private async isAuthorized(args: {
```

Comment:
nitpick: maybe `isValid` is a better description of what we do here 🤔

Authorization seems more related with resource access. A request can be valid (expiration time, nonce) but not authorised to access a resource.

Reply:
Changed in 719b0f301fa3901b5ec468b7c90c04936b592f8c.